Witnesses made visible in FCS

dsyme opened the pull request:
This is an FCS feature requested by @alfonsogarciacaro for Fable and also relevant to WebSharper and others

It surfaces the information about witness passing through the FCS API

This will make F# transpilers more semantically accurate

The key additions can be seen in the signatrue files, e.g. 
```fsharp
type FSharpExpr =
   ....
    /// Like Call but also indicates witness arguments
    val (|CallWithWitnesses|_|) : FSharpExpr -> (FSharpExpr option * FSharpMemberOrFunctionOrValue * FSharpType list * FSharpType list * FSharpExpr list * FSharpExpr list) option 

    /// Indicates a witness argument index from the witness arguments supplied to the enclosing method
    val (|WitnessArg|_|) : FSharpExpr -> int option

type FSharpMemberOrFunctionOrValue = 

    /// Check if this method has an entrpoint that accepts witness arguments and if so return
    /// the name of that entrypoint and the types of the witness arguments
    member GetWitnessPassingInfo: unit -> (string * IList<FSharpType>) option
```

TODO
* [x] testing
* [x] hopefully @alfonsogarciacaro can check if the information is sufficient for Fable

auduchinok commented on src/fsharp/symbols/Symbols.fsi:
```diff
@@ -59,7 +61,8 @@ type [<Class>] public FSharpDisplayContext =
 /// or FSharpActivePatternCase.
-type [<Class>] public FSharpSymbol = 
+[<Class>] 
+type public FSharpSymbol = 
```
Wouldn't it be better to also remove these `public` keywords given the symbols would be public without them anyway?

alfonsogarciacaro commented:
This is great, @dsyme, thanks a lot! I can PR some tests to this branch next week with SRTP samples that have been reported recently as not being resolved correctly by Fable :+1:

KevinRansom commented:
@dsyme , this is marked as wip, is it still a wip?

alfonsogarciacaro commented:
@dsyme I've built FCS from this branch and I'm testing it with Fable: https://github.com/fable-compiler/Fable/blob/1ed140e4ffacf232ca5d14e5bee2e48e5b1c3570/src/Fable.Transforms/FSharp2Fable.fs#L548-L562

It's working but I'm facing a couple of problems:

- The index from `WitnessArg` doesn't seem to be reliable. Every time I've found the pattern is always -1, so I've to make a guess and assume it's 0 if the enclosing scope does contain witnesses.
- When visiting some expressions the following line is throwing the exception: "The lists had different lengths. list2 is 1 element shorter than list1": https://github.com/dotnet/fsharp/blob/ed9b9b7df908bbc2b90e46de27dc05678a6d787a/src/fsharp/MethodCalls.fs#L1864

For example I get the exception when trying to parse any of the following texts (note the message always says "list2 is 1 element shorter than list1"):

```fsharp
type Point =
    { x: int; y: int }
    static member Zero = { x=0; y=0 }
    static member Neg(p: Point) = { x = -p.x; y = -p.y }
    static member (+) (p1, p2) = { x= p1.x + p2.x; y = p1.y + p2.y }

type MyNumber =
    | MyNumber of int
    static member Zero = MyNumber 0
    static member (+) (MyNumber x, MyNumber y) =
        MyNumber(x + y)
    static member DivideByInt (MyNumber x, i: int) =
        MyNumber(x / i)

type MyNumberWrapper =
    { MyNumber: MyNumber }

module ArrayTests =
    testCase "Array.average works with custom types" <| fun () ->
        [|MyNumber 1; MyNumber 2; MyNumber 3|] |> Array.average |> equal (MyNumber 2)

    testCase "Array.averageBy works with custom types" <| fun () ->
        [|{ MyNumber = MyNumber 5 }; { MyNumber = MyNumber 4 }; { MyNumber = MyNumber 3 }|]
        |> Array.averageBy (fun x -> x.MyNumber) |> equal (MyNumber 4)

    testCase "Array.sum with non numeric types works" <| fun () ->
        let p1 = {x=1; y=10}
        let p2 = {x=2; y=20}
        [|p1; p2|] |> Array.sum |> equal {x=3;y=30}

    testCase "Array.sumBy with non numeric types works" <| fun () ->
        let p1 = {x=1; y=10}
        let p2 = {x=2; y=20}
        [|p1; p2|] |> Array.sumBy Point.Neg |> equal {x = -3; y = -30}

    testCase "Array.sum with non numeric types works II" <| fun () ->
        [|MyNumber 1; MyNumber 2; MyNumber 3|] |> Array.sum |> equal (MyNumber 6)

    testCase "Array.sumBy with non numeric types works II" <| fun () ->
        [|{ MyNumber = MyNumber 5 }; { MyNumber = MyNumber 4 }; { MyNumber = MyNumber 3 }|]
        |> Array.sumBy (fun x -> x.MyNumber) |> equal (MyNumber 12)
```

I will try to convert these into tests for this repository and send a PR tomorrow.

dsyme commented:
@alfonsogarciacaro Thank you so much for trialling this, I'll take a look at the problems you mention and get things under test too.

dsyme commented:
> For example I get the exception when trying to parse any of the following texts (note the message always says "list2 is 1 element shorter than list1"):

Do you know if you were using the FSharp.Core from a nuget package (without witness passing) or the freshly built FSharp.Core (with witness passsing)

alfonsogarciacaro commented:
Ah, I had a feeling it might be related to FSharp.Core. I was indeed using the one from Nuget, let me try with a freshly built one.

dsyme commented:
> Ah, I had a feeling it might be related to FSharp.Core. I was indeed using the one from Nuget, let me try with a freshly built one.

Well, we need to be robust when using the nuget one, I was just more asking for my own information when I look into this

alfonsogarciacaro commented:
Hmm, I tried using the `artifacts/bin/FSharp.Core/Proto/netstandard2.0/FSharp.Core.dll` file I got after running `fcs/build.sh` in this branch, both for the Tests project and Fable itself but still getting the same error :/

As a side note, I must say that building this repo in macOS is a breeze compared to what it used to be. Kudos to everybody involved in improving this!

ncave commented:
Adding `--langversion:preview` (see #10140) seems to be breaking FCS tests [here](https://github.com/dotnet/fsharp/blob/feature/fcsw/src/fsharp/symbols/Exprs.fs#L497-L499).

cartermp commented:
FYI @dsyme @alfonsogarciacaro we are planning on release FCS 38 alongside F# 5, which is just in a few short weeks. If this is important to get in by then, we should prioritize addressing all outstanding issues with this one so that it's not a mad rush to publish.

alfonsogarciacaro commented:
For what it concerns Fable, right now we are using a custom build of FCS so we don't depend on the package, so please don't rush things just to let this in, we can build from this repo later (we can even build from the branch). Although other projects like Websharper may do be waiting for this to be included in the FCS package, see https://github.com/dotnet-websharper/core/issues/1113

dsyme commented:
This is now read (once green)

dsyme commented:
@KevinRansom As part of this I addressed #10140 and updated relevant baselines.  These tests are now always run in `--preview` mode to give proper coverage of the witness passing generation (if you like this can be removed once latest release is integrated back into main)

dsyme commented:
@alfonsogarciacaro if you get a chance to check this with Fable I'd appreciate it. Use Fsharp.Core 5.0.0

dsyme commented:
@KevinRansom @TIHan @cartermp Can I have a review on this please?

@ncave This fixes a performance problem with `FSharpExpr` where delayed expressions were continually being re-expanded and never storing their expansion. You might notice a perf improvement in Fable because of this.

alfonsogarciacaro commented:
Thanks a lot for this @dsyme! @ncave used this branch to create the custom build we're using in Fable (it contains a "slim" service we use to improve incremental compilations) and I'm trying it now with Fable tests.

Unfortunately, I'm getting similar errors as before: in a few cases the custom operator cannot be resolved because there are no witnesses in the context (seems to happen in non inlined wrappers, for example [here](https://github.com/fable-compiler/Fable/blob/nagareyama/tests/Main/ApplicativeTests.fs#L78-L79)). And the compilation of tests in Array/List/Seq modules fails again with `The lists had different lengths`.

Not sure if I'm doing something wrong when passing the witnesses, could you please check? I still don't understand what's the -1 index in `BasicPatterns.WitnessArg` so I'm just ignoring these expressions and accumulating the rest of witnesses: https://github.com/fable-compiler/Fable/pull/2240/files#diff-10431e89b8f55bac576b87b3cd8b318e8fca8270a27200cffb0bf0060fe2ecb1R506-R529

alfonsogarciacaro commented:
Sorry, forget about the context with no witnesses, I think for non inlined functions I need to get the witnesses from the member declaration. But the "List had different lengths" is still an issue. It seems to happen when calling BasicPatterns.CallWithWitness in some methods of the collection modules.

dsyme commented:
> But the "List had different lengths" is still an issue. It seems to happen when calling BasicPatterns.CallWithWitness in some methods of the collection modules.

OK thanks.  I'll take a look, likely in a couple of days time

alfonsogarciacaro commented:
@ncave I pushed a branch to reproduce the issue, I hope it helps

- Clone https://github.com/fable-compiler/Fable/
- Checkout branch `witnesses-repro`
- In your IDE, edit the Debug config for `Fable.Cli` project to start in the root dir of the repo and pass `src/Quicktest` as program arguments. See below for an example with Rider.
- Start the debugger, an exception is thrown here: https://github.com/dotnet/fsharp/blob/07e12af88fc84947dabf6b78f0cc139995e63537/src/fsharp/MethodCalls.fs#L1874

`argExprs` has one item, a value of type unit, `argTypes` is empty. This is the code that causes the error:

```fsharp
type Point =
    { x: int; y: int }
    static member Zero = { x=0; y=0 }
    static member (+) (p1, p2) = { x= p1.x + p2.x; y = p1.y + p2.y }

let p1 = {x=1; y=10}
let p2 = {x=2; y=20}
[p1; p2] |> List.sum |> (=) {x=3;y=30} |> printfn "%A"
```

<img width="1069" alt="Screen Shot 2020-10-28 at 15 03 09" src="https://user-images.githubusercontent.com/8275461/97399655-f2595f00-1930-11eb-99a7-15f3e1899b14.png">

alfonsogarciacaro commented:
BTW, if I change the code at that point to swallow the exception and rebuild FCS, then all Fable tests pass :tada: (Except for a few one where I need to adjust how the witnesses are passed through.) Obviously this is not the ideal solution but it seems there are situations where the expected argument types can be ignored (maybe in properties like `get_Zero`).

```fsharp
let convertedArgs =
    try (argExprs, argTypes) ||> List.map2 (fun expr expectedTy -> mkCoerceIfNeeded g expectedTy (tyOfExpr g expr) expr)
    with _ -> argExprs
```

alfonsogarciacaro commented:
@dsyme Now I realize that calling `FSharpMemberOrFunctionOrValue.GetWitnessPassingInfo()` always returns `None`, I guess this is unexpected too? There doesn't seem to be another way to access the witness besides `BasicPatterns.CallWithWitnesses`.

alfonsogarciacaro commented:
I've tried to debug this a bit and what I could find is the `TypedTree.TyVar list` of `vref.Type` in [this line](https://github.com/dotnet/fsharp/pull/9510/files#diff-e2d04f578e35c150aab1e086266bf7a4902baf8d68cfb5cb5743a09751942c79R2111) always have empty `.Constraints` so this function always returns an empty list: https://github.com/dotnet/fsharp/blob/886205d54e27a2e6c0e796bb380b6846b427fe19/src/fsharp/TypedTreeOps.fs#L2309-L2312

alfonsogarciacaro commented:
In contrast, in the `FSharpExprConvert.GetWitnessArgs` call stack there's a point where the constraints of the type parameters are set. Maybe something like that is missing in `.GetWitnessPassingInfo()`?

```
FSharpExprConvert.GetWitnessArgs 
> ConstraintSolver.CodegenWitnessesForTyparInst
> ConstraintSolver.FreshenTypeInst
> ConstraintSolver.FreshenAndFixupTypars
> ConstraintSolver.FixupNewTypars
```
https://github.com/dotnet/fsharp/blob/886205d54e27a2e6c0e796bb380b6846b427fe19/src/fsharp/infos.fs#L381

dsyme commented:
@alfonsogarciacaro Thanks, I'm setting things up to repro your situation

I assume the README in `lib\fcs` in the Fable repro is accurate?

```
C:\GitHub\dsyme\Fable>more lib\fcs\README.md
# FSharp.Compiler.Service

This binary is generated by the following steps:

- Clone https://github.com/ncave/fsharp
- Check out `service_slim` branch
- Run `bash fcs/build.sh`
- Copy `artifacts/bin/FSharp.Compiler.Service/Release/netstandard2.0/FSharp.Compiler.Service.*`
- Copy `artifacts/bin/FSharp.Compiler.Service/Release/netstandard2.0/FSharp.Core.*`
```

dsyme commented:
Actually I guess I want this branch

service_slim_fcsw       -> ncave/service_slim_fcsw

dsyme commented:
@ncave I'm have some trouble building `ncave/service_slim_fcsw`. I get

```
C:\GitHub\dsyme\ncave\src\buildtools\buildtools.targets(30,5): error MSB3073: The command ""C:\Program Files\dotnet\dotnet.exe" "C:\GitHub\dsyme\ncave\artifacts\\bin\fslex\Release\netcoreapp3.1\fslex.dll" -o "C:\GitHub\dsyme\ncave\artifacts\obj\FSharp.Compiler.Private\Proto\net472\illex.fs" --unicode --lexlib Internal.Utilities.Text.Lexing ..\absil\illex.fsl" exited with code 1. [C:\GitHub\dsyme\ncave\src\fsharp\FSharp.Compiler.Private\FSharp.Compiler.Private.fsproj]
```

after running

```
.\build.cmd
```
in that branch.  I'll try various techniques to unblock

alfonsogarciacaro commented:
@dsyme Yes, the instructions are correct. And for the witnesses you are correct it is `ncave/fsharp:service_slim_fcsw` branch. I think you get the error if you run `./build.cm`, you must run `bash fcs/build.sh` (that is, the script in the "fcs" directory) instead.

alfonsogarciacaro commented:
A full script to recompile FCS and test it with Fable, assuming you're in Fable directory and `ncave/fsharp:service_slim_fcsw` is in the "ncave" sibling directory, would be:

```
cd ../ncave
dotnet build -c Release ../ncave/src/buildtools/buildtools.proj
dotnet build -c Release ../ncave/src/fsharp/FSharp.Compiler.Service
cp ../ncave/artifacts/bin/FSharp.Compiler.Service/Release/netstandard2.0/FSharp.Compiler.Service.* lib/fcs/
cp ../ncave/artifacts/bin/FSharp.Compiler.Service/Release/netstandard2.0/FSharp.Core.* lib/fcs/
dotnet run -p src/Fable.Cli -- src/quicktest
```

dsyme commented:
OK, the problem with List.sum repros in main as a regression when taking quotations of code involving List.sum (when trying to generate the quotation data).

@cartermp This is a regression, I presume we will see this in the wild. I will add a separate issue for this.

```fsharp
type Point =
    { x: int; y: int }
    static member Zero = { x=0; y=0 }
    static member (+) (p1, p2) = { x= p1.x + p2.x; y = p1.y + p2.y }
let p1 = {x=1; y=10}
let p2 = {x=2; y=20}

<@ List.sum [p1] @>
```

dsyme commented:
Tracked here: https://github.com/dotnet/fsharp/issues/10364

dsyme commented:
@alfonsogarciacaro I think I'm doing this right, that QuickTest now seems to pass:

```
C:\GitHub\dsyme\Fable>dotnet run -p src/Fable.Cli -- src/quicktest
Fable: F# to JS compiler 3.0.0-nagareyama-beta-004
Thanks to the contributor! @vbfox
src\quicktest> cmd /C "dotnet" restore QuickTest.fsproj
  Determining projects to restore...
  Restored C:\GitHub\dsyme\Fable\src\quicktest\QuickTest.fsproj (in 406 ms).
Parsing src\quicktest\QuickTest.fsproj...
Initializing F# compiler...
Compiling src\quicktest\QuickTest.fsproj...
F# compilation finished in 2454ms
Compiled src\quicktest\QuickTest.fs
Fable compilation finished in 904ms
```

dsyme commented:
This is ready once @alfonsogarciacaro and @ncave  confirm it works for Fable without additional issues

alfonsogarciacaro commented:
This is awesome @dsyme, thanks a lot for all your help! I've merged your latest changes and all Fable tests are passing now :tada: :tada: :tada:

> `GetWitnessPassingInfo` still doesn't seem to work, but the cases where `BasicPatterns.TraitCall` appears without inlining can be resolved by simply selecting the appropriate method using the trait info.

The [performance fix](https://github.com/dotnet/fsharp/pull/9510#issuecomment-717427743) is also very welcome! So far when we were building active patterns by composing `BasicPatterns` (so we were visiting expressions more than once) there was always a performance hit, and this was probably the reason.

The trial by fire will now be to compile FSharpPlus with Fable ;)

dsyme commented:
> This is awesome @dsyme, thanks a lot for all your help! I've merged your latest changes and all Fable tests are passing now 🎉 🎉 🎉

This is great.  This PR is now ready to merge

> The trial by fire will now be to compile FSharpPlus with Fable ;)

Just a word on this: that library is complex and relies on baroque aspects of SRTP resolution that use the feature in ways that have never been recommended F# practice, and which to be honest I believe make for a poor user experience (e.g. in error messages, simplicity and compiler performance, let alone the problem I have with throwing the Functor/Applicative/... computational abstractions at beginners who deeply don't need them). I occasionally use it as a stress test, and use it as a way to isolate out test cases that capture the exact nature of those baroque corners, and it's useful for flushing those out.  It's those individual extracted tests that I consider the acceptance criteria rather.  If it does cause problems, please extract out a minimal test case, but we'll fix it as a separate bug.

dsyme commented:
@KevinRansom @TIHan @cartermp Can I have a review on this please? It's ready to go in.

KevinRansom reviewed (approved):
Thanks for this Don

cartermp reviewed (approved):
Thanks - been following along as this progressed. Lots of formatting changes but they're all for the better.

alfonsogarciacaro commented:
The release candidate of Fable 3 has been published with these changes and it's working flawlessly (at least in my tests. Thanks again for all your work @dsyme!

Also, not entirely sure if it's all due to the performance fix in this PR but I made another simple benchmark of the compiler and the Fable part compilation (after FCS ParseAndCheck) is now around 25% faster! 🚀 This is in the addition to the performance improvements that were already presented at F# eXchange.

forki commented:
Awesome! So exciting
